### DISCUSSION TITLE
`mini.clue` append to triggers after calling `setup` function

### DISCUSSION THREAD

**Discussion by farzadmf** [nvim-mini/mini.nvim#871] [Q&A] [answered]
Hi there,

In my setup, I lazy load all my plugins and some of them are activated using `lazy.nvim`'s `keys` option.

Also, I have set up things in a way that I "append" to my `which-key` config when a plugin is setup/loaded (so that if I disable the plugin, I don't see the "empty" descriptions in `which-key`)

I want to achieve a similar thing with `mini.clue`. I checked the doc and I think all the examples show the configuration along with calling `setup` function.

But for me, I want to do things _after_ my `mini.clue`'s `setup` (which uses `event = 'VeryLazy'`). 

Is there a way I can update `triggers` after `setup`?

**Comment by echasnovski** [comment, marked as answer]
Hi!

This is possible, but not really what 'mini.clue' was designed for. Hence a bit awkward.

Currently used config of all modules is available as `config` field of module's global variable (as per [general principles](https://github.com/echasnovski/mini.nvim?tab=readme-ov-file#general-principles)). So of 'mini.clue' it would be `MiniClue.config`.

So modifying triggers is as easy as directly modifying `MiniClue.config.triggers`. **But** with the following caveats:
- It has to be done *after* `require('mini.clue').setup()`, because it relies on the `MiniClue` global variable.
- Specifically triggers are [implemented as buffer-local mappings](https://github.com/echasnovski/mini.nvim/blob/2d0f10540efdbb1f5c480ebe19ce4de451936a5d/doc/mini-clue.txt#L52). So in order for them to be up to date for already opened buffers, there needs to be a call to either [`ensure_all_triggers()` or `ensure_buf_triggers()`](https://github.com/echasnovski/mini.nvim/blob/2d0f10540efdbb1f5c480ebe19ce4de451936a5d/doc/mini-clue.txt#L610-L620). This would not have been a problem for modifying, say, clues.

So all in all, I guess the answer is "yes, but with caveats because 'mini.clue' is not really designed for this use case". I'd advise setting triggers in the single `require('mini.clue').setup()` and relying on the mapping's descriptions provided by the plugins (which automatically used as clues).

**Reply by farzadmf** [reply]
Thank you @echasnovski for the detailed answer.

The main reason I'm using this method is for something like this: let's say I load a plugin that adds its own "group" of key bindings, so I want that plugin to be in charge; if it loads, it brings in its keys (and their group), if not, no binding is in the list.

If I set up all the groups in `mini.clue`'s `setup`, and that plugin is not loaded (for whatever reason), then I'll see an empty group with no bindings inside. (I use `wk.register(...)` to achieve this.

**Reply by echasnovski** [reply]
> The main reason I'm using this method is for something like this: let's say I load a plugin that adds its own "group" of key bindings, so I want that plugin to be in charge; if it loads, it brings in its keys (and their group), if not, no binding is in the list.

Now, that is usually a bit different than registering triggers. I don't see any particular harm in registering common triggers (like `<Leader>`, `g`, etc.) inside `require('mini.clue').setup()`.

If plugin adds its own mappings, they are picked up automatically based on their description. So here all there is needed to do is add them when making mappings.

Adding the "group" is as straightforward as `table.insert(MiniClue.config.clues, { mode = 'n', keys = '<Leader>f', desc = '+Find' })`. This still requires `require('mini.clue').setup()` to be called prior, but no need to ensure triggers.

**Reply by farzadmf** [reply]
Thank you @echasnovski for your answer. Let me elaborate a bit more maybe it would be more clear:

My use case:
- As I mentioned, I lazy load all my plugins, and I have a plugin that's loading on `keys`
- I have a "generic" group `<leader>c` for coding stuff.
- I want this new plugin to bring in, say, `<leader>cd` for "coding/debug".
- Then, inside of "coding/debug", I'll have individual mappings relted to the plugin.
- Now, I want `<leader>cd` to only be displayed if this specific plugin is loaded because, if it's not, it will be an empty list (which is not very useful).

Trying it now, I tried to print `MiniClue` in:
- `init()`
- `opts` (which I'm setting to a `function()` instead of a static Lua table)
- And also inside `keys` (which is also set to a `function()`).

And in all places, `MiniClue` is `nil` (although this plugin is loaded on `keys`, but `mini.clue` is loaded on `VeryLazy`). I tried to add `mini.clue` in `dependencies`, hoping to get a table for `MiniClue`, but it's still `nil`.

**Reply by echasnovski** [reply]
> And in all places, `MiniClue` is `nil` (although this plugin is loaded on `keys`, but `mini.clue` is loaded on `VeryLazy`). I tried to add `mini.clue` in `dependencies`, hoping to get a table for `MiniClue`, but it's still `nil`.

Sorry, but this is more about how you load plugins. What I can say from 'mini.nvim' point of view, is that `require('mini.clue').setup()` should be executed prior to be able to use `_G.MiniClue` as global variable.

**Comment by farzadmf** [comment]
Hey @echasnovski , a follow-up on this:

As a "real world example" of what I'm hoping to achieve is [this LazyVim commit](https://github.com/LazyVim/LazyVim/commit/d8644c47158fd28b88f5e3a90af6e5e6a0df2eba#diff-e994d682ffb4dd478c1d184131fcde72a9cc6bf08a347c115af449ad71ef328fR63-R110), which, ironically, is done for `mini.ai`, done by a Neovim legend just like yourself, `folke`.

As you can see, some mappings are created dynamically and registered using `require('which-key').register(...)`; if someone sets `enabled = false` for `mini.ai` in there (just for the sake of example), nothing will break and nothing needs updating; just that those mappings/descriptions won't be registered.

Following your guidance above, I managed to register some of my existing "groups" in `mini.clue` (by calling `table.insert(MiniClue.config.clues, ...)`, but trying to adapt the LazyVim commit I mentioned to `mini.clue` doesn't seem to work.

I'm trying to use the `i` and `a` tables defined in there for `mini.clue`, with something like this:
```lua
for k, v in pairs(i) do
  table.insert(MiniClue.config.clues, { mode = 'x', keys = 'i' .. k, desc = v })
  table.insert(MiniClue.config.clues, { mode = 'o', keys = 'i' .. k, desc = v })
  table.insert(MiniClue.config.clues, { mode = 'x', keys = 'a' .. k, desc = v:gsub(' including.*', '') })
  table.insert(MiniClue.config.clues, { mode = 'o', keys = 'a' .. k, desc = v:gsub(' including.*', '') })
end
```

I've also set `v` as a `trigger`:
```lua
triggers = {
  -- ...
  { mode = 'n', keys = 'v' },
},
```

But, doing something like `v` in normal mode, I see this:
![image](https://github.com/echasnovski/mini.nvim/assets/5768792/ae05dead-132e-4907-89b0-3eaa33a69730)

Which feels like none of my `table.insert`s above had any effect.

Also, I think `mini.ai`, [by default](https://github.com/echasnovski/mini.ai/blob/a0262e61f79dfc7bb940d1719c00b0ede042ddd7/lua/mini/ai.lua#L1143-L1146), adds the `Around/inside textobject` descriptions to mappings, which don't seem to be picked up by `mini.clue` and my configured `v` trigger.

Sorry if it became long; tried to show that I'm _actually_ trying things and not asking questions without anything on my side.

**Reply by echasnovski** [reply]
There are several issues with this, which need to be addressed first before investigating further:

- Adding `{ mode = 'n', keys = 'v' }` as trigger is not a good idea. This implies that you want to show *Normal* mode mappings that start with `v`. Instead the goal (I think) is to show all Visual mode mappings. Unfortunately, 'mini.clue' doesn't work like that (show clues "across" modes) and it is a known limitation. The best attempt to show all mappings when mode changed was #442, but it felt to introduce too much complication to the module.
    The best suggestion here is to use `{ mode = 'x', keys = 'i' }` and `{ mode = 'x', keys = 'a' }` to show clues after pressing "i" / "a" in Visual mode.
- Trying out 'mini.clue' features with 'which-key.nvim' enabled is also not a good idea. It seems to be the case here because of that strange symbol present in clues ('which-key.nvim' uses it for its internal purposes). Try completely disabling 'which-key.nvim' and take a look again.
